manifests/server/grant.pp

```diff
@@ -187,37 +239,58 @@
       $schema = $object_name
 
       # Again there seems to be no easy way in plain SQL to check if ALL
-      # PRIVILEGES are granted on a table. By convention we use INSERT
-      # here to represent ALL PRIVILEGES (truly terrible).
-      $custom_privilege = $_privilege ? {
-        'ALL'            => 'INSERT',
-        'ALL PRIVILEGES' => 'INSERT',
-        default          => $_privilege,
+      # PRIVILEGES are granted on a table.
+      # There are currently 7 possible priviliges:
+      # ('SELECT','UPDATE','INSERT','DELETE','TRIGGER','REFERENCES','TRUNCATE')
+      # This list is consistant from Postgresql 8.0
+      #
+      # There are 4 cases to cover, each with it's own distinct unless clause:
+      #    grant ALL
+      #    grant SELECT (or INSERT or DELETE ...)
+      #    revoke ALL
+      #    revoke SELECT (or INSERT or DELETE ...)
+
+      if $ensure == 'present' {
+        if $_privilege == 'ALL' or $_privilege == 'ALL PRIVILEGES' {
+          # GRANT ALL
+          $custom_unless = "SELECT 1 FROM
+             ( SELECT tablename FROM pg_catalog.pg_tables
+               WHERE schemaname = '${schema}' EXCEPT
+                SELECT table_name FROM
+                (SELECT table_name,count(privilege_type) FROM information_schema.role_table_grants
+                  WHERE grantee = '${role}' AND table_schema = '${schema}'
+                  AND privilege_type IN ('SELECT','UPDATE','INSERT','DELETE','TRIGGER','REFERENCES','TRUNCATE')
+                  GROUP BY table_name
+                ) AS P WHERE P.count >= 7
+             ) AS TBLS HAVING TBLS.count=0"
```

Comment:
I would much rather use 'onlyif' instead of 'unless' for these queries, but for 2 things:
* The existing puppet code was written this way. It uses `$unless_function = 'custom'` plus this code to determine whether or not to apply the GRANT. Adding the logic for `$onlyif_function = 'custom'` would make `ALL TABLES IN SCHEMA` the exception to this approach
* The variable `$onlyif_function` is currently only used to test for the existence of something - a table, a language. Re-using it for a different purpose add complexity to this logic.

It's quite feasible to user `onlyif`, but it's really just simplifying the SQL and making the puppet code more complex.

Let me know your thoughts on this trade-off. I'll change it if you ask.

On the rest of the SQL, the logic around `GRANT ALL ON ALL TABLES IN SCHEMA` is quite complicated. I want to capture the case where, say one table has one privilege missing. So it is necessary to account for _all_ privileges, not just any one. Hence the `count(privilege_type)`.
This has to be done on a per-table basis, which is why the `GROUP BY` is there.

count(*) would suffice, I just put privilege_type in there for readability, since that is what we are interested in.

I'm happy to substitute your version.